rfd/0206-stricter-linting-for-the-frontend.md

```diff
@@ -0,0 +1,413 @@
+lead to runtime errors if the value is actually `null` or `undefined`. Instead of using `!`, the code should be
+refactored to
+ensure that the value is not `null` or `undefined` before using it.
+
+### and more...
+
+There are a lot of other rules in ESLint's TypeScript plugin that can help to improve the quality of the codebase. The
+full list [is here](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/src/configs/strict-type-checked.ts).
+
+There are also [stylistic type checked rules](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/src/configs/stylistic-type-checked.ts)
+which can help to improve the consistency and readability of the codebase.
+
+## Migration plan
+
+These rules cannot be enabled all at once, or even one-by-one, as there would be thousands of errors which would be hard
+to fix in one PR and backport, and could also potentially disrupt other pull requests in flight.
+
+Instead, we can move the existing codebase over to a legacy directory in one large swoop, and then enable the strict
+changes for the new directories, left in the same place as the old codebase was. We would then move code over from
+the legacy folder to the new folder, fixing the strict errors as we go. Backporting changes should be relatively easy,
+as the diff to existing code should be small (ideally only import changes to point to the new codebase).
+
+Due to the way the TypeScript compiler works, it is not possible to have strictly typed code import non-strictly typed
+code, without also strictly checking the non-strictly typed code. This means that we cannot have a mix of strict and
+non-strict code in the same directory.
+
+It is possible for the non-strict code to import strictly typed code, so as we migrate parts of the codebase over, the
+existing codebase can import and use the code from the new location.
+
+When moving code over, instead of deleting the legacy file once finished, the file should instead export all the
+exports from the new file, marked with a `@deprecated` comment. This will minimize the disruptions when updating 
+Teleport E, as import changes will not break the E build, as E can still point to the legacy locations.
+
+Once the migration is complete, we would be left with the codebase being strictly typed.
+
+Whilst we introduce a new design component system (outlined
+in [RFD 196](https://github.com/gravitational/teleport/pull/53157)) (something else that would also require the
+separation of code into a new directory during a migration), it
+is a great opportunity to also add strict type checking.
+
+### Current directory structure
+
+```
+- e
+  - web
+    - teleport
+- web
+  - packages
+    - design
+    - shared
+    - teleport
+    - teleterm
+```
+
+### Proposed directory structure
+
+```
+- e
+  - web
+    - new
+    - teleport
+- web
+  - packages
+    - legacy
```

Comment:
It might be easier to append/prepend `legacy` as to existing packages than to put them all in a separate directory called `legacy`. With a separate dir, I'm not sure how pnpm and other tools would handle a configuration like this:

```
packages:
  - web/packages/*
  - web/packages/legacy/*
```

Those tools might not have problems with that at all, but I feel like it might be worthwhile to make the setup as close to a normal setup as possible.